File: src/lib/advisor.ts
/**
 * FPL Transfer Advisor — Strategic Tagging & Recommendation Engine
 * ================================================================
 * TypeScript port of the Python brain's advisory logic.
 * Assigns strategic tags and generates natural-language reasoning
 * for transfer recommendations displayed in the dashboard.
 */

import type { FullElement, PlayerProjection } from "./xpts";

// ---------- Tag Definitions ----------

export type TagId =
  | "template"
  | "differential"
  | "ultra_differential"
  | "trap"
  | "value_beast";

export interface PlayerTag {
  id: TagId;
  label: string;
  emoji: string;
  context: string;
  color: string; // Tailwind text color class
  bgColor: string; // Tailwind bg class for badge
}

export const TAG_DEFS: Record<TagId, PlayerTag> = {
  template: {
    id: "template",
    label: "Template",
    emoji: "🛡️",
    context: "High ownership safety pick.",
    color: "text-blue-400",
    bgColor: "bg-blue-600/20 text-blue-400 border border-blue-500/30",
  },
  differential: {
    id: "differential",
    label: "Differential",
    emoji: "📈",
    context: "Rank climber.",
    color: "text-purple-400",
    bgColor: "bg-purple-600/20 text-purple-400 border border-purple-500/30",
  },
  ultra_differential: {
    id: "ultra_differential",
    label: "Ultra-Diff",
    emoji: "🚀",
    context: "High risk, massive reward potential.",
    color: "text-orange-400",
    bgColor: "bg-orange-600/20 text-orange-400 border border-orange-500/30",
  },
  trap: {
    id: "trap",
    label: "Trap",
    emoji: "⚠️",
    context: "Overperforming stats — likely to regress.",
    color: "text-red-400",
    bgColor: "bg-red-600/20 text-red-400 border border-red-500/30",
  },
  value_beast: {
    id: "value_beast",
    label: "Value",
    emoji: "💰",
    context: "Frees up budget for premiums.",
    color: "text-emerald-400",
    bgColor: "bg-emerald-600/20 text-emerald-400 border border-emerald-500/30",
  },
};

// ---------- Tagged Player ----------

export interface TaggedTarget {
  playerId: number;
  webName: string;
  teamId: number;
  teamShortName: string;
  position: number;
  price: number; // in 0.1m units
  xPts: number; // next GW expected points
  ownership: number; // percentage
  form: number;
  tags: PlayerTag[];
  reasoning: string;
}

export interface TransferPair {
  playerOut: {
    id: number;
    webName: string;
    teamShortName: string;
    price: number;
    xPts: number;
    tags: PlayerTag[];
  };
  playerIn: TaggedTarget;
  xpGain: number;
  budgetDelta: number; // positive = saves money
  reasoning: string;
  priority: number;
}

// ---------- Core Tagging Engine ----------

export function tagPlayer(
  element: FullElement,
  projection: PlayerProjection | undefined,
  allProjections: PlayerProjection[],
): PlayerTag[] {
  const tags: PlayerTag[] = [];
  const ownership = parseFloat(element.selected_by_percent || "0");
  const form = parseFloat(element.form || "0");
  const price = element.now_cost / 10;

  // Get xP for next 3 GWs (using single-GW projection as proxy)
  const xpNext = projection?.expected_points ?? 0;
  const xpNext3 = xpNext * 3; // Approximate

  // Top-25% threshold
  const allXps = allProjections
    .map((p) => p.expected_points)
    .sort((a, b) => b - a);
  const top25Threshold = allXps[Math.floor(allXps.length / 4)] ?? 0;

  // Value ratio: xP per £m over next 3 GWs
  const valueRatio = price > 0 ? xpNext3 / price : 0;

  // --- Template (High Ownership) ---
  if (ownership > 30) {
    tags.push(TAG_DEFS.template);
  }

  // --- Differential ---
  if (ownership < 10 && xpNext > top25Threshold) {
    tags.push(TAG_DEFS.differential);
  }

  // --- Ultra-Differential ---
  if (ownership < 2) {
    tags.push(TAG_DEFS.ultra_differential);
  }

  // --- The Trap ---
  // Compare actual points_per_game to expected output
  const ppg = parseFloat(element.points_per_game || "0");
  const xgPerGame =
    parseFloat(element.expected_goals || "0") / Math.max(element.starts, 1);
  const xaPerGame =
    parseFloat(element.expected_assists || "0") / Math.max(element.starts, 1);
  // Rough expected ppg: appearance(2) + goals*pts + assists*3
  const goalPts = { 1: 6, 2: 6, 3: 5, 4: 4 }[element.element_type] ?? 4;
  const expectedPpg = 2 + xgPerGame * goalPts + xaPerGame * 3;
  if (expectedPpg > 0 && ppg > expectedPpg * 1.5 && element.starts >= 5) {
    tags.push(TAG_DEFS.trap);
  }

  // --- Value Beast ---
  if (valueRatio > 0.8) {
    tags.push(TAG_DEFS.value_beast);
  }

  return tags;
}

// ---------- Build Tagged Transfer Targets ----------

export function buildTaggedTargets(
  projections: PlayerProjection[],
  elements: FullElement[],
  teams: { id: number; short_name: string }[],
  squadIds: number[],
  maxResults: number = 8,
): TaggedTarget[] {
  const elementMap = new Map(elements.map((e) => [e.id, e]));
  const teamMap = new Map(teams.map((t) => [t.id, t.short_name]));
  const squadSet = new Set(squadIds);

  // Filter to non-squad, reliable starters only
  // minutes_probability >= 0.6 AND minimum 5 starts to avoid low-sample inflation
  const candidates = projections.filter((p) => {
    if (squadSet.has(p.player_id)) return false;
    if (p.minutes_probability < 0.6) return false;
    const el = elementMap.get(p.player_id);
    if (!el) return false;
    // Hard filter: must have meaningful game time
    if (el.starts < 5) return false;
    if (el.minutes < 400) return false;
    return true;
  });

  return candidates.slice(0, maxResults).map((proj) => {
    const el = elementMap.get(proj.player_id);
    const tags = el ? tagPlayer(el, proj, projections) : [];
    const teamName = teamMap.get(proj.team_id) ?? "???";
    const ownership = parseFloat(el?.selected_by_percent || "0");
    const form = parseFloat(el?.form || "0");

    // Build reasoning
    const reasonParts: string[] = [];
    if (tags.some((t) => t.id === "differential")) {
      reasonParts.push(`Only ${ownership.toFixed(1)}% owned — rank climber`);
    }
    if (tags.some((t) => t.id === "value_beast")) {
      reasonParts.push(
        `£${((el?.now_cost ?? 0) / 10).toFixed(1)}m — elite value per £`,
      );
    }
    if (tags.some((t) => t.id === "template")) {
      reasonParts.push(`${ownership.toFixed(0)}% owned — must-have`);
    }
    if (form >= 5) {
      reasonParts.push(`Form: ${form.toFixed(1)} — hot streak`);
    }
    if (reasonParts.length === 0) {
      reasonParts.push(`${proj.expected_points.toFixed(1)} xPts projected`);
    }

    return {
      playerId: proj.player_id,
      webName: proj.web_name,
      teamId: proj.team_id,
      teamShortName: teamName,
      position: proj.position,
      price: el?.now_cost ?? 0,
      xPts: proj.expected_points,
      ownership,
      form,
      tags,
      reasoning: reasonParts.join(". ") + ".",
    };
  });
}

// ---------- Build Transfer Pairs (Sell → Buy) ----------

export function buildTransferPairs(
  squadPicks: { element: number; position: number }[],
  projections: PlayerProjection[],
  elements: FullElement[],
  teams: { id: number; short_name: string }[],
  maxPairs: number = 5,
): TransferPair[] {
  const elementMap = new Map(elements.map((e) => [e.id, e]));
  const teamMap = new Map(teams.map((t) => [t.id, t.short_name]));
  const projMap = new Map(projections.map((p) => [p.player_id, p]));
  const squadIds = new Set(squadPicks.map((p) => p.element));

  // Count teams in squad for team-limit check
  const teamCounts = new Map<number, number>();
  for (const pick of squadPicks) {
    const el = elementMap.get(pick.element);
    if (el) {
      teamCounts.set(el.team, (teamCounts.get(el.team) ?? 0) + 1);
    }
  }

  const pairs: TransferPair[] = [];

  for (const pick of squadPicks) {
    const outEl = elementMap.get(pick.element);
    if (!outEl) continue;

    const outProj = projMap.get(pick.element);
    const outXp = outProj?.expected_points ?? 0;
    const outPrice = outEl.now_cost;
    const outTags = outEl ? tagPlayer(outEl, outProj, projections) : [];

    // Find best replacement at same position within budget
    // Hard filter: must have >= 5 starts and >= 400 minutes to avoid low-sample picks
    const candidates = projections.filter((p) => {
      if (squadIds.has(p.player_id)) return false;
      if (p.minutes_probability < 0.6) return false;
      const cEl = elementMap.get(p.player_id);
      if (!cEl) return false;
      if (cEl.element_type !== outEl.element_type) return false;
      if (cEl.starts < 5 || cEl.minutes < 400) return false;
      // Team limit: after removing outEl's team slot, can we add this player?
      const currentTeamCount = teamCounts.get(cEl.team) ?? 0;
      const adjustedCount =
        cEl.team === outEl.team ? currentTeamCount - 1 : currentTeamCount;
      if (adjustedCount >= 3) return false;
      return true;
    });

    for (const inProj of candidates) {
      const inEl = elementMap.get(inProj.player_id);
      if (!inEl) continue;

      const xpGain = inProj.expected_points - outXp;
      if (xpGain <= 0.3) continue; // Only meaningful upgrades

      const budgetDelta = (outPrice - inEl.now_cost) / 10;
      const inTags = tagPlayer(inEl, inProj, projections);

      // Build reasoning
      const reasonParts: string[] = [];
      const pctOutput =
        outXp > 0
          ? Math.round((inProj.expected_points / outXp) * 100)
          : 100;
      if (pctOutput >= 100) {
        reasonParts.push(
          `${inEl.web_name} projects ${inProj.expected_points.toFixed(1)} xPts vs ${outEl.web_name}'s ${outXp.toFixed(1)}`,
        );
      } else if (pctOutput >= 85) {
        reasonParts.push(
          `${inEl.web_name} provides ${pctOutput}% of ${outEl.web_name}'s output`,
        );
      }

      if (budgetDelta > 0.5) {
        reasonParts.push(`saving £${budgetDelta.toFixed(1)}m to upgrade elsewhere`);
      }
      if (inTags.some((t) => t.id === "differential")) {
        reasonParts.push("differential pick for rank gains");
      }
      if (inTags.some((t) => t.id === "value_beast")) {
        reasonParts.push("elite value per £m");
      }
      if (outTags.some((t) => t.id === "trap")) {
        reasonParts.push(`${outEl.web_name} flagged as regression risk`);
      }

      pairs.push({
        playerOut: {
          id: pick.element,
          webName: outEl.web_name,
          teamShortName: teamMap.get(outEl.team) ?? "???",
          price: outPrice,
          xPts: outXp,
          tags: outTags,
        },
        playerIn: {
          playerId: inProj.player_id,
          webName: inEl.web_name,
          teamId: inEl.team,
          teamShortName: teamMap.get(inEl.team) ?? "???",
          position: inEl.element_type,
          price: inEl.now_cost,
          xPts: inProj.expected_points,
          ownership: parseFloat(inEl.selected_by_percent || "0"),
          form: parseFloat(inEl.form || "0"),
          tags: inTags,
          reasoning: "",
        },
        xpGain: Math.round(xpGain * 10) / 10,
        budgetDelta: Math.round(budgetDelta * 10) / 10,
        reasoning:
          reasonParts.length > 0
            ? reasonParts.join(". ") + "."
            : "Fixture-based upgrade.",
        priority: 0,
      });
    }
  }

  // Sort by xP gain, assign priority
  pairs.sort((a, b) => b.xpGain - a.xpGain);
  return pairs.slice(0, maxPairs).map((p, i) => ({ ...p, priority: i + 1 }));
}

// ---------- Safety Score ----------

/**
 * GW-performance based safety score (0–100).
 * 50 = exactly average, scales ±25 per standard deviation.
 * Uses FPL typical std dev of ~13 points.
 */
export function calculateSafetyScore(
  livePoints: number,
  averageScore: number,
): number {
  const STD_DEV = 13;
  const zScore = (livePoints - averageScore) / STD_DEV;
  const raw = 50 + zScore * 25;
  return Math.min(99, Math.max(1, Math.round(raw)));
}

File: brain/advisor.py
"""
FPL Transfer Advisor — Strategic Tagging & Recommendation Engine
=================================================================
Post-processes optimizer output to assign strategic tags and generate
human-readable transfer advice with reasoning.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional, Tuple

from optimizer import (
    OptimizationResult,
    Player,
    SquadState,
    TransferSuggestion,
)


# ---------- Tag Definitions ----------


class Tag(Enum):
    TEMPLATE = "Template"
    DIFFERENTIAL = "Differential"
    ULTRA_DIFFERENTIAL = "Ultra-Differential"
    TRAP = "Trap"
    VALUE_BEAST = "Value Beast"

    @property
    def emoji(self) -> str:
        return {
            Tag.TEMPLATE: "🛡️",
            Tag.DIFFERENTIAL: "📈",
            Tag.ULTRA_DIFFERENTIAL: "🚀",
            Tag.TRAP: "⚠️",
            Tag.VALUE_BEAST: "💰",
        }[self]

    @property
    def context(self) -> str:
        return {
            Tag.TEMPLATE: "High ownership safety pick.",
            Tag.DIFFERENTIAL: "Rank climber.",
            Tag.ULTRA_DIFFERENTIAL: "High risk, massive reward potential.",
            Tag.TRAP: "Overperforming stats significantly—likely to regress.",
            Tag.VALUE_BEAST: "Frees up budget for premiums.",
        }[self]


# ---------- Data Structures ----------


@dataclass
class TaggedPlayer:
    player: Player
    tags: List[Tag]
    tag_reasons: Dict[Tag, str]
    xp_next_3: float = 0.0
    value_ratio: float = 0.0


@dataclass
class TransferAdvice:
    player_out: TaggedPlayer
    player_in: TaggedPlayer
    xp_gain: float
    budget_delta: float  # in £m (positive = saves money)
    reasoning: str
    priority: int  # 1 = highest


@dataclass
class AdvisorReport:
    transfer_advice: List[TransferAdvice]
    captain_pick: TaggedPlayer
    vice_captain_pick: TaggedPlayer
    squad_warnings: List[str]
    summary: str


# ---------- Core Tagging Engine ----------


def assign_player_tags(
    player: Player,
    all_players: List[Player],
    horizon_gws: List[int],
) -> TaggedPlayer:
    """
    Assign strategic tags to a player based on ownership, form,
    expected points, and value metrics.
    """

    tags: List[Tag] = []
    reasons: Dict[Tag, str] = {}

    # Calculate aggregate stats
    xp_next_3 = sum(player.xp_by_gw.get(gw, 0.0) for gw in horizon_gws[:3])
    price_m = player.now_cost / 10.0
    value_ratio = xp_next_3 / price_m if price_m > 0 else 0.0

    # Calculate top-25% xP threshold from all players
    all_xp_next_3 = sorted(
        [sum(p.xp_by_gw.get(gw, 0.0) for gw in horizon_gws[:3]) for p in all_players],
        reverse=True,
    )
    top_25_threshold = all_xp_next_3[len(all_xp_next_3) // 4] if all_xp_next_3 else 0.0

    # --- Tag 1: Template (High Ownership) ---
    if player.ownership_pct > 30.0:
        tags.append(Tag.TEMPLATE)
        reasons[Tag.TEMPLATE] = (
            f"Owned by {player.ownership_pct:.1f}% of managers. "
            f"Benching risks a red arrow."
        )

    # --- Tag 2: Differential ---
    if player.ownership_pct < 10.0 and xp_next_3 > top_25_threshold:
        tags.append(Tag.DIFFERENTIAL)
        reasons[Tag.DIFFERENTIAL] = (
            f"Only {player.ownership_pct:.1f}% ownership but {xp_next_3:.1f} xP "
            f"over next 3 GWs (top 25% threshold: {top_25_threshold:.1f})."
        )

    # --- Tag 3: Ultra-Differential ---
    if player.ownership_pct < 2.0:
        tags.append(Tag.ULTRA_DIFFERENTIAL)
        reasons[Tag.ULTRA_DIFFERENTIAL] = (
            f"Only {player.ownership_pct:.1f}% ownership. "
            f"Massive rank swing potential if they haul."
        )

    # --- Tag 4: The Trap ---
    if player.xp_last_5 > 0 and player.actual_points_last_5 > (player.xp_last_5 * 1.5):
        overperformance = (
            (player.actual_points_last_5 / player.xp_last_5 - 1) * 100
        )
        tags.append(Tag.TRAP)
        reasons[Tag.TRAP] = (
            f"Scored {player.actual_points_last_5:.0f} pts vs "
            f"{player.xp_last_5:.1f} xP over last 5 GWs "
            f"({overperformance:.0f}% overperformance). Regression likely."
        )

    # --- Tag 5: Value Beast ---
    if value_ratio > 0.8:
        tags.append(Tag.VALUE_BEAST)
        reasons[Tag.VALUE_BEAST] = (
            f"{xp_next_3:.1f} xP / £{price_m:.1f}m = {value_ratio:.2f} value ratio. "
            f"Frees budget for premium picks."
        )

    return TaggedPlayer(
        player=player,
        tags=tags,
        tag_reasons=reasons,
        xp_next_3=round(xp_next_3, 1),
        value_ratio=round(value_ratio, 2),
    )


# ---------- Advice Generator ----------


def generate_transfer_advice(
    suggestions: List[TransferSuggestion],
    all_players: List[Player],
    current_gw: int,
    horizon: int = 5,
) -> List[TransferAdvice]:
    """
    Takes raw transfer suggestions and produces tagged, reasoned advice.
    """
    horizon_gws = list(range(current_gw + 1, current_gw + horizon + 1))
    advice_list: List[TransferAdvice] = []

    for idx, suggestion in enumerate(suggestions):
        tagged_out = assign_player_tags(suggestion.player_out, all_players, horizon_gws)
        tagged_in = assign_player_tags(suggestion.player_in, all_players, horizon_gws)

        budget_delta = (suggestion.player_out.selling_price - suggestion.player_in.now_cost) / 10.0
        reasoning = _build_reasoning(tagged_out, tagged_in, suggestion.xp_gain, budget_delta)

        advice_list.append(
            TransferAdvice(
                player_out=tagged_out,
                player_in=tagged_in,
                xp_gain=suggestion.xp_gain,
                budget_delta=budget_delta,
                reasoning=reasoning,
                priority=idx + 1,
            )
        )

    return advice_list


def generate_full_report(
    result: OptimizationResult,
    suggestions: List[TransferSuggestion],
    all_players: List[Player],
    current_squad: SquadState,
) -> AdvisorReport:
    """
    Generate a full advisory report from optimization results.
    """
    horizon_gws = list(
        range(current_squad.gameweek + 1, current_squad.gameweek + result.horizon + 1)
    )

    # Tag all squad players
    tagged_squad = [
        assign_player_tags(p, all_players, horizon_gws)
        for p in result.suggested_squad
    ]

    # Transfer advice
    advice = generate_transfer_advice(
        suggestions, all_players, current_squad.gameweek, result.horizon
    )

    # Captain & VC
    captain_player = next(
        (p for p in result.suggested_squad if p.id == result.captain_id),
        result.suggested_squad[0],
    )
    vc_player = next(
        (p for p in result.suggested_squad if p.id == result.vice_captain_id),
        result.suggested_squad[1],
    )
    tagged_captain = assign_player_tags(captain_player, all_players, horizon_gws)
    tagged_vc = assign_player_tags(vc_player, all_players, horizon_gws)

    # Squad warnings
    warnings = _detect_squad_warnings(tagged_squad, current_squad)

    # Summary
    summary = _build_summary(result, advice, tagged_captain)

    return AdvisorReport(
        transfer_advice=advice,
        captain_pick=tagged_captain,
        vice_captain_pick=tagged_vc,
        squad_warnings=warnings,
        summary=summary,
    )


# ---------- Internal Helpers ----------


def _build_reasoning(
    out: TaggedPlayer,
    in_player: TaggedPlayer,
    xp_gain: float,
    budget_delta: float,
) -> str:
    """Build a natural language reasoning string for a transfer."""
    parts: List[str] = []

    # xP comparison
    if out.xp_next_3 > 0:
        pct = (in_player.xp_next_3 / out.xp_next_3 * 100) if out.xp_next_3 > 0 else 0
        if pct >= 90:
            parts.append(
                f"{in_player.player.web_name} provides {pct:.0f}% of "
                f"{out.player.web_name}'s output"
            )
        else:
            parts.append(
                f"{in_player.player.web_name} projects {in_player.xp_next_3:.1f} xP "
                f"vs {out.player.web_name}'s {out.xp_next_3:.1f} xP (next 3 GWs)"
            )

    # Budget impact
    if budget_delta > 0.5:
        parts.append(f"saving £{budget_delta:.1f}m to upgrade elsewhere")
    elif budget_delta < -0.5:
        parts.append(f"costs £{abs(budget_delta):.1f}m more but justifies the premium")

    # Fixture advantage
    if in_player.xp_next_3 > out.xp_next_3:
        parts.append("better upcoming fixture run")

    # Tag-based insights
    if Tag.TRAP in out.tags:
        parts.append(f"{out.player.web_name} is flagged as a Trap (regression risk)")
    if Tag.DIFFERENTIAL in in_player.tags:
        parts.append(f"{in_player.player.web_name} is a Differential pick for rank gains")
    if Tag.VALUE_BEAST in in_player.tags:
        parts.append(f"{in_player.player.web_name} offers elite value per £m")

    return ". ".join(parts) + "." if parts else "Marginal improvement based on fixture swing."


def _detect_squad_warnings(
    tagged_squad: List[TaggedPlayer],
    squad_state: SquadState,
) -> List[str]:
    """Detect potential issues in the squad."""
    warnings: List[str] = []

    # Check for traps in starting XI
    traps = [tp for tp in tagged_squad if Tag.TRAP in tp.tags]
    if traps:
        names = ", ".join(tp.player.web_name for tp in traps)
        warnings.append(
            f"Regression alert: {names} flagged as Trap(s). "
            f"Consider selling before price drops."
        )

    # Check template coverage
    templates = [tp for tp in tagged_squad if Tag.TEMPLATE in tp.tags]
    if len(templates) < 3:
        warnings.append(
            f"Low template coverage ({len(templates)} template players). "
            f"Risk of falling behind the crowd."
        )

    # Check for too many players from one team
    team_counts: Dict[int, int] = {}
    for tp in tagged_squad:
        team_counts[tp.player.team_id] = team_counts.get(tp.player.team_id, 0) + 1
    heavy_teams = {tid: c for tid, c in team_counts.items() if c >= 3}
    if heavy_teams:
        warnings.append(
            f"Fixture dependency: {len(heavy_teams)} team(s) with 3 players. "
            f"Blank GW risk."
        )

    # Check for injured/doubtful starters
    flagged = [
        tp for tp in tagged_squad
        if tp.player.status in ("d", "i") and tp.player.form > 3.0
    ]
    if flagged:
        names = ", ".join(tp.player.web_name for tp in flagged)
        warnings.append(f"Fitness concern: {names} flagged as doubtful/injured.")

    return warnings


def _build_summary(
    result: OptimizationResult,
    advice: List[TransferAdvice],
    captain: TaggedPlayer,
) -> str:
    """Build the executive summary string."""
    parts: List[str] = []

    if result.transfers_in:
        transfer_lines = []
        for a in advice:
            tags_str = " ".join(
                f"[{t.emoji} {t.value}]" for t in a.player_in.tags
            )
            transfer_lines.append(
                f"  Sell {a.player_out.player.web_name} → "
                f"Buy {a.player_in.player.web_name} "
                f"(+{a.xp_gain:.1f} xP) {tags_str}"
            )
        parts.append("Recommended transfers:\n" + "\n".join(transfer_lines))
    else:
        parts.append("Recommendation: Roll the transfer. No clear upgrades available.")

    if result.transfer_cost > 0:
        parts.append(
            f"Transfer cost: -{result.transfer_cost} pts "
            f"({len(result.transfers_in) - 1} hit(s)). "
            f"Net gain: {result.net_xp:.1f} xP over {result.horizon} GWs."
        )

    captain_tags = " ".join(f"[{t.emoji} {t.value}]" for t in captain.tags)
    parts.append(
        f"Captain: {captain.player.web_name} "
        f"({captain.xp_next_3:.1f} xP next 3 GWs) {captain_tags}"
    )

    return "\n".join(parts)


# ---------- Formatting Utilities ----------


def format_tags(tags: List[Tag]) -> str:
    """Format tags for display."""
    return " ".join(f"[{t.emoji} {t.value}]" for t in tags)


def format_transfer_card(advice: TransferAdvice) -> str:
    """Format a single transfer suggestion as a display card."""
    out_tags = format_tags(advice.player_out.tags)
    in_tags = format_tags(advice.player_in.tags)
    budget_str = (
        f"+£{advice.budget_delta:.1f}m saved"
        if advice.budget_delta > 0
        else f"-£{abs(advice.budget_delta):.1f}m"
        if advice.budget_delta < 0
        else "Budget neutral"
    )

    return (
        f"#{advice.priority} | "
        f"Sell {advice.player_out.player.web_name} {out_tags} → "
        f"Buy {advice.player_in.player.web_name} {in_tags}\n"
        f"   xP Gain: +{advice.xp_gain:.1f} | {budget_str}\n"
        f"   {advice.reasoning}"
    )
